modules/build/src/main/scala/scala/build/options/BuildOptions.scala

```diff
@@ -273,81 +268,225 @@ final case class BuildOptions(
+    either {
+      import coursier.Versions
+      import coursier.core.Latest
+      import coursier._
+
+      val moduleVersion: Either[ScalaVersionError, String] = {
+        def scala3 = mod"org.scala-lang:scala3-library_3"
+        val res = finalCache.logger.use {
+          Versions(finalCache)
+            .withModule(scala3)
+            .result()
+            .unsafeRun()(finalCache.ec)
+        }
+        res.versions.latest(Latest.Release) match {
+          case Some(versionString) => Right(versionString)
+          case None                => Left(new NetworkUnaccessibleScalaVersionError(None))
+        }
+      }
+
+      val scalaVersion       = value(moduleVersion)
+      val scalaBinaryVersion = ScalaVersion.binary(scalaVersion)
+      (scalaVersion, scalaBinaryVersion)
+    }
+
+  /** @return
+    *   Either a BuildException or the calculated (ScalaVersion, ScalaBinaryVersion) tuple
+    */
+  private def computeLatestScalaTwoNightlyVersions(): Either[BuildException, (String, String)] =
+    either {
+      import coursier.Versions
+      import coursier.core.Latest
+      import coursier._
+
+      val moduleVersion: Either[ScalaVersionError, String] = {
+        def scalaNightly2Module: Module = mod"org.scala-lang:scala-library"
+        val res = finalCache.logger.use {
+          Versions(finalCache)
+            .withModule(scalaNightly2Module)
+            .withRepositories(Seq(coursier.Repositories.scalaIntegration))
+            .result()
+            .unsafeRun()(finalCache.ec)
+        }
+        res.versions.latest(Latest.Release) match {
+          case Some(versionString) => Right(versionString)
+          case None                => Left(new NetworkUnaccessibleScalaVersionError(None))
+        }
+      }
+
+      val scalaVersion       = value(moduleVersion)
+      val scalaBinaryVersion = ScalaVersion.binary(scalaVersion)
+      (scalaVersion, scalaBinaryVersion)
     }
-    val maybeSv = scalaVersion match {
-      case None => matchNewestScalaVersion(None)
-      case Some(sv0) =>
-        if (Util.isFullScalaVersion(sv0)) Right(sv0)
-        else matchNewestScalaVersion(Some(sv0))
+
+  private def turnScala2NightlyVersionArgToVersions(versionString: String)
+    : Either[BuildException, (String, String)] = either {
+
+    val moduleVersion: Either[ScalaVersionError, String] = {
+      import coursier._
+      def scalaNightly2Module: Module = mod"org.scala-lang:scala-library"
+      val res = finalCache.logger.use {
+        Versions(finalCache)
+          .withModule(scalaNightly2Module)
+          .withRepositories(Seq(coursier.Repositories.scalaIntegration))
+          .result()
+          .unsafeRun()(finalCache.ec)
+      }
+      res.versions.available.find(versionString == _) match {
+        case Some(vStr) => Right(vStr)
+        case None => Left(new NoValidScalaVersionFoundError(
+            res.versions.available,
+            latestSupportedStableVersions
+          ))
+      }
```

Comment:
More idiomatic / simple way of doing that:
```suggestion
      if (res.versions.available.contains(versionString)) versionString
      else
        Left(
          new NoValidScalaVersionFoundError(res.versions.available, latestSupportedStableVersions)
        )
```